src/main/proto/netflix/titus/titus_cluster_membership_api.proto

```diff
@@ -0,0 +1,205 @@
+}
+
+message ClusterMembershipRevision {
+
+    ClusterMember current = 1;
+
+    // (Required) Detailed code for being in a specific state.
+    string code = 2;
+
+    // (Optional) Textual description accompanying the 'reasonCode'.
+    string message = 3;
+
+    // Time when a transition to a state happened.
+    uint64 timestamp = 4;
+}
+
+message ClusterMembershipEvent {
+
+    message Snapshot {
+        repeated ClusterMembershipRevision revisions = 1;
+    }
+
+    message Updated {
+        ClusterMembershipRevision revision = 1;
+    }
+
+    message Removed {
+        ClusterMembershipRevision revision = 1;
+    }
+
+    oneof Event {
+        Snapshot snapshot = 1;
+        Updated updated = 2;
+        Removed removed = 3;
+    }
+}
+
+// ----------------------------------------------------------------------------
+// GRPC services
+
+message MemberId {
+    string id = 1;
+}
+
+message ClusterMembers {
+    repeated ClusterMember members = 1;
+}
+
+message QueryRequest {
+    // Filtering criteria:
+    // * 'ids' - comma separated list of instance ids
+    // * 'instanceLimit' - maximum number of records to return per instance
+    // * 'limit' - maximum number of records to return
+    // * 'before' - events before the given timestamp (epoch)
+    // * 'since' - events at or after the given timestamp (epoch)
+    map<string, string> filteringCriteria = 1;
+
+    /// (Optional) If set, only field values explicitly specified in this parameter will be returned.
+    repeated string fields = 2;
+}
+
+message ClusterMembershipRevisions {
+    repeated ClusterMembershipRevision revisions = 1;
+}
+
+message UpdateMemberLabelsRequest {
+    string memberId = 1;
+
+    map<string, string> labels = 2;
+}
+
+message DeleteMemberLabelsRequest {
+    string memberId = 1;
+    repeated string keys = 2;
+}
+
+message UpdateMemberStateRequest {
+    string memberId = 1;
+
+    string reasonMessage = 2;
+
+    ClusterMember.State newState = 3;
+}
+
+service ClusterMembershipService {
+
+    // Get all known cluster members.
+    rpc GetMembers (google.protobuf.Empty) returns (ClusterMembers) {
+    }
+
+    // Get member with the given id.
+    rpc GetMember (MemberId) returns (ClusterMember) {
+    }
+
+    // Get recent changes.
+    rpc GetMembershipRevisions (QueryRequest) returns (ClusterMembershipRevisions) {
+    }
+
+    // Adds all labels from the request object to the target member. Labels that exist are
+    // overridden. Returns the updated object.
+    rpc UpdateMemberLabels (UpdateMemberLabelsRequest) returns (ClusterMember) {
+    }
+
+    // Removes all specified labels from the target object. Labels that do not exist are ignored.
+    // Returns the updated object.
+    rpc DeleteMemberLabels (DeleteMemberLabelsRequest) returns (ClusterMember) {
+    }
+
+    // Change state of a given member. Returns the updated object.
+    rpc UpdateMemberState (UpdateMemberStateRequest) returns (ClusterMember) {
```

Comment:
do we need to provide any consistency guarantees here? How are clients supposed to avoid race conditions when writing to multiple nodes?

Reply:
Ultimately clients requests will be serialized. We cannot provide other guarantees.